### PR TITLE
ARROW-10816: [Rust][DF] Operations with Intervals

### PR DESCRIPTION
**Pull request by ovr**
It's draft, only started to work on it. DRAFT is needed to indicate that I started to work on it.

### CODE REVIEW

**Comment by Dandandan** [comment]
@ovr cool :) could you make a new JIRA story for this PR?

**Comment by ovr** [comment]
@Dandandan I dont think that I should create a new jira issue, because I didnt implement calculation between timestamp and interval.

Btw:

There is a problem, DF executes `BinaryExpr` by casting left and right sides which is not needed for `timestamp - interval`, because this calculation is exception in this case and should be done without casting. I am thinking how it's better to resolve it.

There are two different ideas how to solve it:

1. Rewrite execution for BinaryExpr
2. Wrap operations that dont need cast by function on logical plan

File: rust/arrow/src/datatypes/interval.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::{ArrowNumericType, ArrowPrimitiveType, DataType, IntervalUnit};
+use crate::{make_numeric_type, make_type};
+
+make_type!(
+    IntervalYearMonthType,
+    i32,
```
Comment (jorgecarleitao) [inline rust/arrow/src/datatypes/interval.rs:23]: fwiw, I do not think this is correct and I opened a [issue for this](https://issues.apache.org/jira/browse/ARROW-11645). I think that this should be `struct(i32,i32)` with C representation or something like that.

As it stands, we will be trying to read an `i32` offseted by 4 bytes from an `i64`, which is asking us to write `unsafe` in a lot of places.

These numbers are not something that suits Rust or anything; they _**must**_ be set according to the Arrow specification.

So, I think that we should hold of this PR until we have the interval implementation correct. In general, I am against adding anything to DataFusion until we have integration tests for the type.

Comment (ovr) [reply rust/arrow/src/datatypes/interval.rs:23]: @jorgecarleitao You are right, it's will be better.

**Comment by ovr** [comment]
@alamb @andygrove @nevi-me @jorgecarleitao 

> Btw:

There is a problem, DF executes `BinaryExpr` by casting left and right sides which is not needed for `timestamp - interval`, because this calculation is exception in this case and should be done without casting. I am thinking how it's better to resolve it.

There are two different ideas how to solve it:

1. Rewrite execution for BinaryExpr
2. Wrap operations that dont need cast by function on logical plan

What is better to use? Thanks

**Comment by alamb** [comment]
> There is a problem, DF executes BinaryExpr by casting left and right sides which is not needed for timestamp - interval, because this calculation is exception in this case and should be done without casting. I am thinking how it's better to resolve it.

I think the "casts" in this case are very cheap -- namely no data is actually copied

I am thinking for example `Int32` -> `Time32` cast https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/cast.rs#L597

 calls `cast_array_data` - https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/cast.rs#L871-L885 which basically just re-uses the same underlying data and reinterprets it as the desired type

So I am not sure if there is a major problem or not to have a cast if the cast is effectively doing no real work

**Comment by alamb** [comment]
The Apache Arrow Rust community is moving the Rust implementation into its own dedicated github repositories [arrow-rs](https://github.com/apache/arrow-rs) and [arrow-datafusion](https://github.com/apache/arrow-datafusion). It is likely we will not merge this PR into this repository

Please see the [mailing-list](https://lists.apache.org/thread.html/rce7c61cb3f703367dc00984530016e3fcb828e5a8035655fbcccf862%40%3Cdev.arrow.apache.org%3E) thread for more details

We expect the process to take a few days and will follow up with a migration plan for the in-flight PRs.

**Comment by alamb** [comment]
https://github.com/apache/arrow/pull/10096 has removed the arrow implementation from this repository (it now resides in https://github.com/apache/arrow-rs and https://github.com/apache/arrow-datafusion) in the hopes of streamlining the development process

Please re-target this PR (let us know if you need help doing so) to one/both of the new repositories. 

Thank you for understanding and helping to make arrow-rs and datafusion better

**Pull request closed without merging**